lib/cretonne/src/ir/frontend.rs

```diff
@@ -0,0 +1,764 @@
+pub struct FuncInstBuilder<'short, 'long: 'short, Variable: 'long>
+    where Variable: EntityRef + Hash + Default
+{
+    builder: &'short mut FunctionBuilder<'long, Variable>,
+    ebb: Ebb,
+}
+
+impl<'short, 'long, Variable> FuncInstBuilder<'short, 'long, Variable>
+    where Variable: EntityRef + Hash + Default
+{
+    fn new<'s, 'l>(builder: &'s mut FunctionBuilder<'l, Variable>,
+                   ebb: Ebb)
+                   -> FuncInstBuilder<'s, 'l, Variable> {
+        FuncInstBuilder { builder, ebb }
+    }
+}
+
+impl<'short, 'long, Variable> InstBuilderBase<'short> for FuncInstBuilder<'short, 'long, Variable>
+    where Variable: EntityRef + Hash + Default
+{
+    fn data_flow_graph(&self) -> &DataFlowGraph {
+        &self.builder.func.dfg
+    }
+
+    fn data_flow_graph_mut(&mut self) -> &mut DataFlowGraph {
+        &mut self.builder.func.dfg
+    }
+
+    // This implementation is richer than `InsertBuilder` because we use the data of the
+    // instruction being inserted to add related info to the DFG and the SSA building system,
+    /// and perform debug sanity checks.
+    fn build(self, data: InstructionData, ctrl_typevar: Type) -> (Inst, &'short mut DataFlowGraph) {
+        if data.opcode().is_return() {
+            self.builder
+                .check_return_args(data.arguments(&self.builder.func.dfg.value_lists))
+        }
+        // We only insert the Ebb in the layout when an instruction is added to it
+        if self.builder.position.pristine {
+            self.builder
+                .func
+                .layout
+                .append_ebb(self.builder.position.ebb);
+            self.builder.position.pristine = false;
+        } else {
+            debug_assert!(!self.builder.builder.ebbs[self.builder.position.ebb].filled,
+                          "you cannot add an instruction to a block already filled");
+        }
+        let inst = self.builder.func.dfg.make_inst(data.clone());
+        self.builder.func.dfg.make_inst_results(inst, ctrl_typevar);
+        {
+            let mut cur = Cursor::new(&mut self.builder.func.layout);
+            cur.goto_bottom(self.ebb);
+            cur.insert_inst(inst);
+        }
+        if data.opcode().is_branch() {
+            match data.branch_destination() {
+                Some(dest_ebb) => {
+                    self.builder.declare_successor(dest_ebb, inst);
+                    // If the user has supplied jump arguments we must adapt the arguments of
+                    // the destination ebb
+                    // TODO: find a way not to allocate a vector
+                    let ty_to_append: Option<Vec<Type>> =
+                        match data.analyze_branch(&self.builder.func.dfg.value_lists) {
+                            BranchInfo::SingleDest(_, args) => {
+                                let dest_ebb_args = self.builder.func.dfg.ebb_args(dest_ebb);
+                                if args.len() != 0 {
```

Comment:
OK so here the situation is complicated. Indeed, there are two types of Ebb arguments: those supplied by the user and those produced by the SSA construction. User supplied args are added to the Ebb's arg list when the first jump instruction to this Ebb is inserted. SSA arguments are added to the Ebb's arg list during the insertion of instructions to this Ebb.

That said, here's how the algorithm should work when you insert a jump instruction targeting `ebb`:
- If `ebb` already has a predecessor, then the user supplied args of the instruction being inserted should match the user supplied args of the other jump instructions having `ebb` as dest;
- if `ebb` has no predecessors, then the jump args of thie instruction being inserted are added as user supplied arguments to `ebb`. 

Your comment made me realize that my implementation has a bug: indeed, SSA produced `Ebb` arguments can be mixed with user supplied arguments in the Ebb's arg list because SSA undef values (when the block is not sealed) are directly inserted in the Ebb's arg list instead of being left in a detached state.

I will then use your suggestion to simplify this code in `frontend.rs` and also fix the bug in `ssa.rs`.